metagraph/src/graph/representation/succinct/boss.cpp

```diff
@@ -3147,76 +3138,54 @@ void BOSS::call_kmers(Call<edge_index, const std::string&> callback) const {
     }
 }
 
-void BOSS::index_suffix_ranges(size_t suffix_length) {
+void BOSS::index_suffix_ranges(size_t suffix_length, size_t num_threads) {
     assert(suffix_length <= k_);
 
     indexed_suffix_length_ = suffix_length;
-    indexed_suffix_ranges_.clear();
+    indexed_suffix_ranges_ = decltype(indexed_suffix_ranges_)();
 
     if (indexed_suffix_length_ == 0u)
         return;
 
     if (indexed_suffix_length_ * log2(alph_size - 1) >= 64)
         throw std::runtime_error("ERROR: Trying to index too long suffixes");
 
-    std::vector<std::tuple<uint64_t, edge_index, edge_index>> suffix_ranges;
-
     // first, take empty suffix and the entire range of nodes in the BOSS table
-    uint64_t num_suffixes = 1;
-    suffix_ranges.emplace_back(0, 1, W_->size() - 1);
-
-    // grow the suffix length up to |suffix_length - 1|
-    for (size_t len = 1; len < indexed_suffix_length_; ++len) {
-        std::vector<std::tuple<uint64_t, edge_index, edge_index>> narrowed;
-        narrowed.reserve(suffix_ranges.size() * (alph_size - 1));
-
-        for (const auto &[idx, rl, ru] : suffix_ranges) {
+    // will store pairs: begin[0], end[0], begin[1], end[1], ...
+    std::vector<edge_index> suffix_ranges { 1, W_->size() };
+
+    // grow the suffix length up to |suffix_length|
+    for (size_t len = 1; len <= indexed_suffix_length_; ++len) {
+        std::vector<edge_index> narrowed(suffix_ranges.size() * (alph_size - 1), 1);
+        const size_t N = suffix_ranges.size() / 2;
+
+        #pragma omp parallel for num_threads(num_threads)
+        for (size_t i = 0; i < N; ++i) {
+            edge_index rl = suffix_ranges[2 * i];
+            edge_index ru = suffix_ranges[2 * i + 1] - 1;
             // prepend the suffix with one of the |alph_size - 1| possible characters
             for (TAlphabet c = 1; c < alph_size; ++c) {
                 edge_index rl_next = rl;
                 edge_index ru_next = ru;
                 if (!tighten_range(&rl_next, &ru_next, c))
                     continue;
 
-                assert(idx < num_suffixes);
-                narrowed.emplace_back(num_suffixes * (c - 1) + idx,
-                                      rl_next, ru_next);
+                narrowed[2 * i + 2 * (N * (c - 1))] = rl_next;
+                narrowed[2 * i + 2 * (N * (c - 1)) + 1] = ru_next + 1;
```

Comment:
If `tighten_range` in a previous iteration fails, doesn't this lead to extra `rank_W` calls being made since `tighten_range` is called on `(1, 1)`?

Reply:
if it fails, `(rl, ru) = (1, 0)`, so `tighten_range` will skip the computation and will return false.

Reply:
but if `tighten_range` fails, it doesn't modify the values:
https://github.com/ratschlab/metagraph/blob/6a2f4e024bdaa0e8cc549898ee7cad673375c3ec/metagraph/src/graph/representation/succinct/boss.hpp#L674

Reply:
It does ru = ... - 1 in one of the lines above, so ru becomes < rl.